addons/l10n_in/models/product_template.py

```diff
@@ -11,12 +10,12 @@ class ProductTemplate(models.Model):
     l10n_in_hsn_code = fields.Char(string="HSN/SAC Code", help="Harmonized System Nomenclature/Services Accounting Code")
     l10n_in_hsn_description = fields.Char(string="HSN/SAC Description", help="HSN/SAC description is required if HSN/SAC code is not provided.")
 
-    @api.constrains('l10n_in_hsn_code')
-    def _check_hsn_code_validation(self):
+    @api.onchange('l10n_in_hsn_code')
+    def _onchange_l10n_in_hsn_code(self):
         for record in self:
-            company = record.company_id or self.env.company
-            minimum_hsn_len = company.l10n_in_hsn_code_digit
-            check_hsn = record.l10n_in_hsn_code and minimum_hsn_len
-            if check_hsn and len(record.l10n_in_hsn_code) < int(minimum_hsn_len):
-                error_message = _("As per your HSN/SAC code validation, minimum %s digits HSN/SAC code is required.", minimum_hsn_len)
-                raise ValidationError(error_message)
+            active_hsn_code_digit_len = self.env.companies.mapped('l10n_in_hsn_code_digit')
+            # Remove False value if any company have False value in l10n_in_hsn_code_digit
+            active_hsn_code_digit_len = list(filter(bool, active_hsn_code_digit_len))
+            check_hsn = record.sale_ok and record.l10n_in_hsn_code and active_hsn_code_digit_len
+            if check_hsn and len(record.l10n_in_hsn_code) < int(max(active_hsn_code_digit_len)):
+                return {'warning': {'title': "Warning", 'message': _("As per your HSN/SAC code validation, minimum %s digits HSN/SAC code is required.", max(active_hsn_code_digit_len))}}
```

Comment:
@ajjo-odoo Could you please consider removing this message.  It will stop the user unnecessarily while entering data.

Reply:
> @ajjo-odoo Please consider removing this message. It will stop the user unnecessarily while entering data.

i am in strong opinion is to
guide user on master data than on transactional data
here , product form is my master and invoice is transactional,
once master is correct, less chance on invoice of getting warning.